File: apps/mobile/lib/features/property/domain/entities/room_space.dart
import 'package:equatable/equatable.dart';

import 'catalog_types.dart';

/// Espacio individual de la distribución desglosada del inmueble.
///
/// Cada espacio declara sus dimensiones en metros (largo × ancho).
/// El área se deriva como `lengthM * widthM`. Obligatorio para
/// dormitorios (SOP Fase 1), recomendado para el resto de espacios.
class RoomSpace extends Equatable {
  final RoomSpaceType type;
  final String label;
  final double lengthM;
  final double widthM;
  final String? capacityNotes;

  const RoomSpace({
    required this.type,
    required this.label,
    required this.lengthM,
    required this.widthM,
    this.capacityNotes,
  });

  double get areaM2 => lengthM * widthM;

  RoomSpace copyWith({
    RoomSpaceType? type,
    String? label,
    double? lengthM,
    double? widthM,
    String? capacityNotes,
  }) {
    return RoomSpace(
      type: type ?? this.type,
      label: label ?? this.label,
      lengthM: lengthM ?? this.lengthM,
      widthM: widthM ?? this.widthM,
      capacityNotes: capacityNotes ?? this.capacityNotes,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type.name,
      'label': label,
      'length_m': lengthM,
      'width_m': widthM,
      'area_m2': areaM2,
      'capacity_notes': capacityNotes,
    };
  }

  factory RoomSpace.fromJson(Map<String, dynamic> json) {
    return RoomSpace(
      type: RoomSpaceType.values.byName(json['type'] as String),
      label: json['label'] as String,
      lengthM: (json['length_m'] as num).toDouble(),
      widthM: (json['width_m'] as num).toDouble(),
      capacityNotes: json['capacity_notes'] as String?,
    );
  }

  @override
  List<Object?> get props => [type, label, lengthM, widthM, capacityNotes];
}

File: apps/backend/src/modules/property/domain/entities/room-space.ts
import { RoomSpaceType } from './catalog-types';

/**
 * Espacio individual de la distribución desglosada del inmueble.
 *
 * Espejo TypeScript de `RoomSpace` en mobile. Cada espacio declara sus
 * dimensiones en metros (largo × ancho); el área se deriva.
 * Obligatorio para dormitorios (SOP Fase 1), recomendado para el resto.
 */
export interface RoomSpaceProps {
  type: RoomSpaceType;
  label: string;
  lengthM: number;
  widthM: number;
  capacityNotes?: string | null;
}

export class RoomSpace {
  readonly type: RoomSpaceType;
  readonly label: string;
  readonly lengthM: number;
  readonly widthM: number;
  readonly capacityNotes: string | null;

  private constructor(props: Required<RoomSpaceProps>) {
    this.type = props.type;
    this.label = props.label;
    this.lengthM = props.lengthM;
    this.widthM = props.widthM;
    this.capacityNotes = props.capacityNotes;
    Object.freeze(this);
  }

  static create(props: RoomSpaceProps): RoomSpace {
    return new RoomSpace({
      capacityNotes: null,
      ...props,
    });
  }

  get areaM2(): number {
    return this.lengthM * this.widthM;
  }

  /** Equivalente a `copyWith` en mobile. */
  copy(patch: Partial<RoomSpaceProps>): RoomSpace {
    return RoomSpace.create({
      type: patch.type ?? this.type,
      label: patch.label ?? this.label,
      lengthM: patch.lengthM ?? this.lengthM,
      widthM: patch.widthM ?? this.widthM,
      capacityNotes: patch.capacityNotes ?? this.capacityNotes,
    });
  }

  toJSON(): Record<string, unknown> {
    return {
      type: this.type,
      label: this.label,
      length_m: this.lengthM,
      width_m: this.widthM,
      area_m2: this.areaM2,
      capacity_notes: this.capacityNotes,
    };
  }

  static fromJSON(json: Record<string, unknown>): RoomSpace {
    const type = json.type as RoomSpaceType;
    if (!Object.values(RoomSpaceType).includes(type)) {
      throw new Error(`Unknown RoomSpaceType: ${String(json.type)}`);
    }
    return RoomSpace.create({
      type,
      label: json.label as string,
      lengthM: (json.length_m as number) ?? 0,
      widthM: (json.width_m as number) ?? 0,
      capacityNotes: (json.capacity_notes as string | null) ?? null,
    });
  }

  equals(other: RoomSpace): boolean {
    return (
      this.type === other.type &&
      this.label === other.label &&
      this.lengthM === other.lengthM &&
      this.widthM === other.widthM &&
      this.capacityNotes === other.capacityNotes
    );
  }
}
